store/src/java/com/zimbra/cs/milter/MilterHandler.java

```diff
@@ -354,6 +355,33 @@ private void SMFIC_Rcpt() throws ServiceException {
             connection.send(new MilterPacket(SMFIR_TEMPFAIL));
             return;
         }
+        try { // ZBUG-2165: prevent forwarding to non-permitted distribution list
+            ZimbraLog.milter.trace("evaluating forwarding to non-permitted distribution list");
+            final Account account = prov.getAccountByName(rcpt);
+            if (account != null) {
+                ZimbraLog.milter.trace("account: %s", account.getName());
+                final String prefMailForwardingAddress = account.getPrefMailForwardingAddress();
+                ZimbraLog.milter.trace("prefMailForwardingAddress: %s", prefMailForwardingAddress);
+                if (prefMailForwardingAddress != null) {
+                    final String[] addresses = prefMailForwardingAddress.split(",");
+                    for (final String a : addresses) {
+                        final String address = a.trim();
+                        ZimbraLog.milter.trace("address: %s", address);
+                        if (!address.isEmpty() && prov.isDistributionList(address)) {
+                            final Group group = prov.getGroupBasic(Key.DistributionListBy.name, address);
+                            ZimbraLog.milter.trace("group: %s", group);
+                            if (group != null && !accessMgr.canDo(account, group, User.R_sendToDistList, false)) {
+                                ZimbraLog.milter.debug("Recipient %s is not allowed to forward to this distribution list: %s", account.getName(), group.getName());
+                                SMFIR_ReplyCode("571", "571 Recipient " + account.getName() + " is not allowed to forward to this distribution list: " + group.getName());
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
There are lots of nested if please refactor this code and try to make it simpler and more readable.
Also, make sure you put only risky code inside a `try` `catch` block.

Reply:
Done.

Refactored code for simplicity and readability.